packages/router/utils.ts

```diff
@@ -1002,9 +1000,12 @@ function compilePath(
   return [matcher, params];
 }
 
-function safelyDecodeURI(value: string) {
+function decodePath(value: string) {
   try {
-    return decodeURI(value);
+    return value
+      .split("/")
+      .map((v) => decodeURIComponent(v).replace(/\//g, "%2F"))
+      .join("/");
```

Comment:
`decodeURI` here is a bit too aggressive since it assumes a full URI which may have query params and hash values in it.  Since we're only operating on a pathname, we should instead decode each component individually in case they have pre-encoded hashes, percent signs, etc. in them.

However, if there was an encoded slash - preserve that as `%2F` since we can't introduce new segments via this decoding